src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,1005 @@
+                {
+                    Headers = new List<ChainedHeader>(headers),
+                    Id = jobId
+                });
+
+                this.logger.LogDebug("{0} blocks were requested from puller. Job ID {1} was created.", headers.Count, jobId);
+                this.processQueuesSignal.Set();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Loop that assigns download jobs to the peers.</summary>
+        private async Task AssignerLoopAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellationSource.IsCancellationRequested)
+            {
+                try
+                {
+                    await this.processQueuesSignal.WaitAsync(this.cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                    return;
+                }
+
+                await this.AssignDownloadJobsAsync().ConfigureAwait(false);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Loop that continuously checks if peers failed to deliver important blocks in given time and penalizes them if they did.</summary>
+        private async Task StallingLoopAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellationSource.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(StallingLoopIntervalMs, this.cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                    return;
+                }
+                
+                this.CheckStalling();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Assigns downloads from <see cref="reassignedJobsQueue"/> and <see cref="downloadJobsQueue"/> to the peers that are capable of delivering blocks.</summary>
+        private async Task AssignDownloadJobsAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var failedHashes = new List<uint256>();
+            var newAssignments = new List<AssignedDownload>();
+
+            lock (this.queueLock)
+            {
+                // First process reassign queue ignoring slots limitations.
+                this.ProcessQueueLocked(this.reassignedJobsQueue, newAssignments, failedHashes);
+
+                // Process regular queue.
+                int emptySlots;
+                lock (this.assignedLock)
+                {
+                    emptySlots = this.maxBlocksBeingDownloaded - this.assignedDownloadsByHash.Count;
+                }
+
+                this.logger.LogTrace("There are {0} empty slots.", emptySlots);
```

Comment:
This trace could go outside the lock as well

Reply:
But we need to log it after we calculated amount of slots and before we process the queue. So I'm not sure if we can move it out of the lock

Reply:
General rule try to log outside locks!

Reply:
Fixed